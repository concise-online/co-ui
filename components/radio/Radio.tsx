import React, { Component, createRef, DOMAttributes, CSSProperties, ChangeEvent, RefObject, ReactText } from 'react';
import classNames from 'classnames';
import RadioGroup from './RadioGroup';
import { runCallback } from '../_utils/function';
import { getPrefixCls } from '../_utils/config';

const prefixCls = getPrefixCls('radio');

export interface Props extends Omit<DOMAttributes<HTMLInputElement>, 'onChange'> {
  value?: any;
  name?: string;
  checked?: boolean;
  disabled?: boolean;
  children?: ReactText,
  className?: string;
  style?: CSSProperties;
  onChange?: (value: any, e: ChangeEvent<HTMLInputElement>) => void;
}

class Radio extends Component<Props> {
  static Group = RadioGroup;
  static isRadio = true;
  private readonly ref: RefObject<HTMLInputElement>;
  constructor(props: Props) {
    super(props);
    this.ref = createRef();
    this.handleChange = this.handleChange.bind(this);
  }

  focus() {
    (this.ref.current as HTMLInputElement).focus();
  }

  blur() {
    (this.ref.current as HTMLInputElement).blur();
  }

  handleChange(e: ChangeEvent<HTMLInputElement>) {
    const { onChange, disabled, value } = this.props;

    if (disabled) return;
    runCallback(onChange, value, e);
  }

  render() {
    const { children, className, checked, name, disabled, style, ...radioProps } = this.props;
    const classes = classNames(prefixCls, {
      [`${prefixCls}-checked`]: checked,
      [`${prefixCls}-disabled`]: disabled,
    }, className);

    return (
      <label className={classes} style={style}>
        <span className={`${prefixCls}-element`}>
          <input
            {...radioProps}
            name={name}
            disabled={disabled}
            checked={checked}
            type="radio"
            onChange={this.handleChange}
            ref={this.ref}
          />
        </span>
        <span className={`${prefixCls}-text`}>
          {children}
        </span>
      </label>
    );
  }
}

export default Radio;
