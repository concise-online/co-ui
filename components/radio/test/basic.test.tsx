import React from 'react';
import { render, mount } from 'enzyme';
import Radio from '../Radio';

describe('radio basic feature', () => {
  it('renders correctly', () => {
    const wrapper = render(<Radio>Radio</Radio>);
    const wrapper1 = render(<Radio checked />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper1).toMatchSnapshot();
  });

  it('change event correctly', () => {
    const container = document.createElement('div');
    document.body.appendChild(container);
    const onChange = jest.fn();
    const wrapper = mount((
      <Radio onChange={onChange}>Radio</Radio>
    ), { attachTo: container });

    wrapper.find('input').simulate('change');
    expect(onChange).toBeCalled();
  });

  it('focus method', () => {
    const container = document.createElement('div');
    document.body.appendChild(container);
    const onFocus = jest.fn();
    const wrapper = mount(<Radio onFocus={onFocus} />, { attachTo: container });

    (wrapper.instance() as Radio).focus();
    expect(onFocus).toBeCalled();
  });

  it('blur method', () => {
    const container = document.createElement('div');
    document.body.appendChild(container);
    const onBlur = jest.fn();
    const wrapper = mount(<Radio onBlur={onBlur} />, { attachTo: container });

    (wrapper.instance() as Radio).focus();
    (wrapper.instance() as Radio).blur();
    expect(onBlur).toBeCalled();
  });
});
