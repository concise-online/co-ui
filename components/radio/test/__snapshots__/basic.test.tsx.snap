// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`radio basic feature renders correctly 1`] = `
<label
  class="co-radio"
>
  <span
    class="co-radio-element"
  >
    <input
      type="radio"
    />
    <span
      class="co-radio-inner"
    />
  </span>
  <span
    class="co-radio-text"
  >
    Radio
  </span>
</label>
`;

exports[`radio basic feature renders correctly 2`] = `
<label
  class="co-radio co-radio-checked"
>
  <span
    class="co-radio-element"
  >
    <input
      checked=""
      type="radio"
    />
    <span
      class="co-radio-inner"
    />
  </span>
  <span
    class="co-radio-text"
  />
</label>
`;
